src/Equinox.MessageDb/MessageDbClient.fs

```diff
@@ -0,0 +1,126 @@
+
+                let timelineEvent =
+                    TimelineEvent.Create(
+                        index = reader.GetInt64(0),
+                        eventType = reader.GetString(1),
+                        data = JsonSerializer.Deserialize<JsonElement>(reader.GetString(2)),
+                        meta = JsonSerializer.Deserialize<JsonElement>(meta),
+                        eventId = reader.GetGuid(4),
+                        ?correlationId =
+                            (if reader.IsDBNull(5) then
+                                 None
+                             else
+                                 Some(reader.GetString(5))),
+                        ?causationId =
+                            (if reader.IsDBNull(6) then
+                                 None
+                             else
+                                 Some(reader.GetString(6))),
+                        timestamp = timestamp
+                    )
+
+                events.Add(timelineEvent)
+                let! next = reader.ReadAsync(ct)
+                hasNext <- next
+
+            return events.ToArray()
+        }
+
+    member private _.PrepareWriteCommand
+        (cmd: NpgsqlBatchCommand)
+        (streamName: string)
+        version
+        (message: IEventData<JsonElement>)
+        =
+        cmd.CommandText <- "select 1 from write_message(@Id::text, @StreamName, @EventType, @Data, @Meta, @ExpectedVersion)"
+
+        cmd.Parameters.AddWithValue("Id", NpgsqlDbType.Uuid, message.EventId) |> ignore
+        cmd.Parameters.AddWithValue("StreamName", NpgsqlDbType.Text, streamName) |> ignore
+        cmd.Parameters.AddWithValue("EventType", NpgsqlDbType.Text, message.EventType) |> ignore
+        let data, meta = message.Data, message.Meta
+        cmd.Parameters.AddWithValue("Data", NpgsqlDbType.Jsonb, data.GetRawText()) |> ignore
+        let meta =
+            match meta.ValueKind with
+            | JsonValueKind.Null | JsonValueKind.Undefined -> "null"
+            | _ -> meta.GetRawText()
+        cmd.Parameters.AddWithValue("Meta", NpgsqlDbType.Jsonb, meta) |> ignore
+        cmd.Parameters.AddWithValue("ExpectedVersion", NpgsqlDbType.Bigint, version) |> ignore
+
+        cmd
+
+    member __.WriteMessages(streamName, events, version: int64, ct) =
+        task {
+            try
+                use! conn = source ()
+                use transaction = conn.BeginTransaction()
+                use batch = new NpgsqlBatch(conn, transaction)
+
+                let mutable expectedVersion = version
+
+                for event in events do
+                    batch.BatchCommands.Add(
+                        __.PrepareWriteCommand (NpgsqlBatchCommand()) streamName expectedVersion event
+                    )
+
+                    expectedVersion <- expectedVersion + 1L
+
+                do! batch.ExecuteNonQueryAsync(ct) :> Task
+                do! transaction.CommitAsync(ct)
+            with :? PostgresException as ex ->
```

Comment:
Equinox eventually wants this to be a DU anyway so maybe map it as soon as possible like the `.EventStore` impl does and avoid having a WEV exception type in here at all. (in `.EventStoreDb`, the current impl uses the API variant that doesnt throw IIRC)
Failing that...>
L125 `raise ex` destroys the stack trace so should be either 
- `reraise` to avoid
- use a `when` clause on the with so it's not trapped in the first instance.